### ISSUE TITLE
Progress Bar epilepsy when progress bar forces terminal scrolling

### ISSUE THREAD

**Issue by tedivm** [tqdm/tqdm#677]
This is building off of the same project from #670 (a multithreaded uploader/downloader). 

In it people are occasionally uploading as many as several hundred files. Everything works great right up until the system goes beyond the number of rows in the terminal- for example, if the user has 50 rows and the 51 progress bar would cause a scrolling event. Once that occurs the whole application goes into seizure mode, throwing progress bars into random locations and forcing an epic amount of scrolling.

**Comment by tedivm** [comment]
To clarify, it seems that the error occurs when the first progress bar (n=1) gets forced off screen.

**Comment by casperdcl** [comment]
yes, don't know what solution is possible seeing as there physically isn't enough room... `position=i%50`?

**Comment by casperdcl** [comment]
side note: I don't know why you'd try to download >10 things simultaneously - surely there won't be any speedup

**Comment by tedivm** [comment]
It's actually downloading eight at a time, and when one finishes the next starts- 

```python

def download_files_threaded(files):
    p = Pool(8, initializer=tqdm.set_lock, initargs=(RLock(),))
    res = []
    i = 1
    for filename, filedata in files.items():
        if os.path.exists(filename):
            os.remove(filename)
        if parallel_downloads:
            res.append(p.apply_async(download_file, args=(filename, filedata['id'], i)))
        else:
            res.append(p.apply(download_file, args=(filename, filedata['id'], i)))

        i += 1
        # Without this pause some async functions fail.
        time.sleep(0.01)

    [x.wait() for x in res]
    # Other threads output isn't tracked by this thread, so add newlines to move below the progress bars.
    click.echo("\n" * (len(res) - 1))


def download_file(filename, fileid, position=0):
    if os.path.exists(filename):
        os.remove(filename)

    download_url = _make_get('api/files/%s/download' % (fileid))
    r = requests.get(download_url, stream=True)
    r.raise_for_status()
    total_length = int(r.headers.get('content-length'))

    if os.sep in filename:
        path = filename[:filename.rindex(os.sep)]
        if not os.path.exists(path):
            os.makedirs(path)

    chunk_size = 1024
    with open(filename, 'wb') as f, tqdm(
        total=total_length,
        unit_scale=True,
        unit='b',
        desc='Downloading file %s' % (filename),
        position=position) as pbar:

        for chunk in r.iter_content(chunk_size=chunk_size):
            if chunk: # filter out keep-alive new chunks
                f.write(chunk)
                pbar.update(chunk_size)

        # Pause before returning to give progress bar time to write.
        time.sleep(0.01)

```


It works beautifully at first- even if the terminal is fully filled when the command runs, it generally forces the scroll and looks great. It's not until the first progress bar leaves the screen (which generally happens after it's been finished anyways, due to it downloading first) that things blow up.

**Comment by casperdcl** [comment]
- `position=1 + i%8`
- just after `for chunk in ...` before the `with ... tqdm(...) as pbar` block exits, add `with tqdm.get_lock(): print(pbar)`

this potentially means multiple bars with the same position, but at least they all get printed separately upon completion.

**Comment by tedivm** [comment]
It took some effort but I got a demo of this working. This demo even makes sure that no two active process bars are in the same position-

```python
from multiprocessing import Pool, freeze_support, RLock, Lock, Array, Manager
import random
from tqdm import tqdm
import time


def test_bar(id, position_holders):
    total = random.randint(50,125)
    position = get_position(position_holders)
    with tqdm(
        total=total,
        unit_scale=False,
        unit='b',
        desc='Position #%s' % (id),
        position=position) as pbar:

        for i in range(0, total):
            pbar.update(1)
            time.sleep(random.randint(1,10)/100)
    release_position(position_holders, position)


def get_position(position_holders):
    position_lock.acquire()
    position_index = list(position_holders).index(0)
    position_holders[position_index] = 1
    position_lock.release()
    return position_index+1

def release_position(position_holders, position):
    position_lock.acquire()
    position_holders[position-1] = 0
    position_lock.release()


def init(tqdm_lock, p_lock):
    tqdm.set_lock(tqdm_lock)
    global position_lock
    position_lock = p_lock


if __name__ == '__main__':
    maxrows = 10
    m = Manager()
    position_holders = m.Array('i', [0] * maxrows)

    p = Pool(8, initializer=init, initargs=(RLock(), m.Lock()))
    res = []

    for i in range(1, 100):
        res.append(p.apply_async(test_bar, args=(i, position_holders)))
        time.sleep(0.01)

    [x.wait() for x in res]
    # Other threads output isn't tracked by this thread, so add newlines to move below the progress bars.

    print("\n" * maxrows)

```

**Comment by casperdcl** [comment]
why do you need 2 locks?

**Comment by tedivm** [comment]
Because the lock for tqdm is stored as a [private class object](https://github.com/tqdm/tqdm/blob/master/tqdm/_tqdm.py#L558), so I can't grab it out of the tqdm class itself. I suppose I could have just created one lock and passed it to both of them, but since we're locking on different types of things anyways (tqdm is locking on output, I'm locking on shared memory access) I felt that this would be more efficient (since memory access wouldn't get blocked by process output).

**Comment by tedivm** [comment]
This also looks a lot better when turning `leave=False` and adding a small delay after the progress bar has disappeared. I've updated the demo code accordingly.

If you're interested I can clean this up and toss it in your examples folder.

**Comment by casperdcl** [comment]
always happy to accept examples & docs

**Comment by casperdcl** [comment]
Fixed by #924

**Comment by tedivm** [comment]
Awesome! Thanks so much!
